nmostesting/nmosTestingComplexCompare.py

```diff
@@ -0,0 +1,262 @@
+    if len(response_constraints_enum) != 1 or len(sample_rate_constraints_enum) != 1:
+        return False
+
+    try:
+        response_numerator = response_constraints_enum[0]["numerator"]
+        response_denominator = 1
+
+        if "denominator" in response_constraints_enum[0]:
+            response_denominator = response_constraints_enum[0]["denominator"]
+
+        sample_rate_numerator = sample_rate_constraints_enum[0]["numerator"]
+        sample_rate_denominator = 1
+
+        if "denominator" in sample_rate_constraints_enum[0]:
+            sample_rate_denominator = sample_rate_constraints_enum[0]["denominator"]
+
+        if (
+            response_numerator != sample_rate_numerator
+            or response_denominator != sample_rate_denominator
+        ):
+            return False
+    except Exception:
+        return False
+
+    # There must be no other patternProperties
+    for prop in sample_rate_constraints["constraint_sets"][0]:
+        if (
+            prop != "urn:x-nmos:cap:format:sample_rate"
+            and prop != "urn:x-nmos:cap:meta:enabled"
+            and prop != "urn:x-nmos:cap:meta:preference"
+        ):
+            return False
+
+    for prop in response_constraints["constraint_sets"][0]:
+        if (
+            prop != "urn:x-nmos:cap:format:sample_rate"
+            and prop != "urn:x-nmos:cap:meta:enabled"
+            and prop != "urn:x-nmos:cap:meta:preference"
+        ):
+            return False
+
+    # Check meta:enabled considering default values
+    response_enabled = True
+    if "urn:x-nmos:cap:meta:enabled" in response_constraints["constraint_sets"][0]:
+        response_enabled = response_constraints["constraint_sets"][0][
+            "urn:x-nmos:cap:meta:enabled"
+        ]
+
+    sample_rate_enabled = True
+    if "urn:x-nmos:cap:meta:enabled" in sample_rate_constraints["constraint_sets"][0]:
+        sample_rate_enabled = sample_rate_constraints["constraint_sets"][0][
+            "urn:x-nmos:cap:meta:enabled"
+        ]
+
+    if response_enabled != sample_rate_enabled:
+        return False
+
+    # Check meta:preference considering default values
+    response_preference = 0
+    if "urn:x-nmos:cap:meta:preference" in response_constraints["constraint_sets"][0]:
+        response_preference = response_constraints["constraint_sets"][0][
+            "urn:x-nmos:cap:meta:preference"
+        ]
+
+    sample_rate_preference = 0
+    if (
+        "urn:x-nmos:cap:meta:preference"
+        in sample_rate_constraints["constraint_sets"][0]
+    ):
+        sample_rate_preference = sample_rate_constraints["constraint_sets"][0][
+            "urn:x-nmos:cap:meta:preference"
+        ]
+
+    if response_preference != sample_rate_preference:
+        return False
+
+    # If we get here it is because the two objects are equal
+    return True
+
+
+def compare_complex_grain_rate_constraint(response_constraints, grain_rate_constraints):
```

Comment:
It seems to me there is a simpler way of comparing `constraint_sets` arrays that is more generic.

1. Iterate through the constraint sets (python [dicts](https://docs.python.org/3/tutorial/datastructures.html#dictionaries)), adding explicit `meta:preference` and `meta:enabled` if not already present.
2. Maybe remove `meta:label` from each constraint set dict if you don't care about that?
4. Insert each constraint set dict into a python [set](https://docs.python.org/3/tutorial/datastructures.html#sets).
5. Use built-in equality comparison on these sets of dicts. (A finesse would be to replace the `enum` arrays also with sets, for the same reason, to avoid dependency on ordering.)

This won't identify all equivalent `constraint_sets`, but many more.

Reply:
The complexity is that the comparison MUST be aware of all the schemas involved ... here constraint_set and rational mainly but who know about the future. Having a function that creates an object from JSON and schema would be required but it seems that this does not exist in the schema supporting library. This is why we created our own, specifically targeted at the need. We could keep a TODO to make this operation more generic but for now we decided to take fastest approach.

Reply:
> The comparison MUST be aware of all the schemas involved

Why?

Reply:
If the schema indicate an optional attribute with default value ... comparing two json objects require knowing that because obj1 may not have the attribute while obj2 has it at the default value ... Both object are identical but you need to know the schema to assess that, simply comparing the JSON does not work.

This is what I meant by "MUST be aware of all the schemas involved"

Reply:
Right, but the only defaults in BCP-004-01 v1.0 are `meta:enabled` and `meta:preference`.

Reply:
https://github.com/AMWA-TV/bcp-004-01/blob/3fb9c19815e63231aa54c35576ec63c3437c72e2/APIs/schemas/param_constraint_rational.json

is another example.

Reply:
Thanks, you're right. Proper rational equality requires more than just handling the default denominator, and I still expect generic BCP-004-01 constraint sets equality can be done in similar number of lines of code. I had better put my own effort in and try it 🙂

Reply:
@gwgeorgea, @alabou, I have quickly had a go at a function to check two Constraint Sets arrays are closely equivalent in the way I outlined. Done during the workshop this morning, so not fully tested yet, but only 42 lines of code, so see what you think:

https://github.com/garethsb/nmos-testing/commit/bec94bcdcec3117d4fcd8c356d33054cb6ab8e82

Reply:
We are now using your class